astropy/io/fits/file.py

```diff
@@ -119,9 +120,16 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
         if mode not in IO_FITS_MODES:
             raise ValueError("Mode '{}' not recognized".format(mode))
 
+        # Handle raw URLs
         if (isinstance(fileobj, string_types) and
-            mode not in ('ostream', 'append') and
-            _is_url(fileobj)):  # This is an URL.
+            mode not in ('ostream', 'append', 'update') and _is_url(fileobj)):
+            self.name = download_file(fileobj, cache=cache)
+        # Handle responses from URL requests that have already been opened
+        elif isinstance(fileobj, http_client.HTTPResponse):
```

Comment:
I added support for `Request` objects. After a little more investigation, I think that even though calling `download_file` opens another resource, it's still a fairly clean implementation. Any alternative would have to store the contents of the `HTTPRequest` to a temporary file and open that, so it would still involve opening and closing other resources. Or it would require a fairly substantial refactor of `download_file`, which might actually be the cleanest solution.